stan/math/rev/core/vari.hpp

```diff
@@ -184,6 +186,400 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. For a sparse eigen matrix
+   * this includes the nozero values as well the inner and outer indices.
+   * Variables should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A sparse Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  explicit vari_value(S&& x)
+      : rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()),
+        val_(x),
+        adj_(x),
+        chainable_alloc() {
+    this->set_zero_adjoint();
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+  /**
+   * Construct an sparse Eigen variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A sparse Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  vari_value(S&& x, bool stacked)
+      : rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()),
+        val_(x),
+        adj_(x),
+        chainable_alloc() {
+    this->set_zero_adjoint();
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return rows_; }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return cols_; }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return size_; }
+  void chain() {}
+  /**
+   * Initialize the adjoint for this (dependent) variable to 1.
+   * This operation is applied to the dependent variable before
+   * propagating derivatives, setting the derivative of the
+   * result with respect to itself to be 1.
+   */
+  inline void init_dependent() {
```

Comment:
Same as above. I think we should just not allow `.grad()` and `.init_dependent()` for these types.